### ISSUE TITLE
Error automatically migrating schema from 7 -> 17

### ISSUE THREAD

**Issue by pjninnis** [hangfire-postgres/Hangfire.PostgreSql#305]
We've just upgraded our application that relies heavily on Hangfire and the Postgres addon.
It's a fairly old application that is still on ASP.NET / NET Framework 4.8

Things were fine with the hangfire database migration in each of our environments until we went into production.

We are now getting a timeout error (see below) when hangfire attempts the migration.
Looks like it's attempting the following: ALTER TABLE "counter" ALTER COLUMN value TYPE bigint

We have upgraded from/to the following packages:

Hangfire v1.6.6 to v1.7.32
Hangfire.Core v1.6.6 to v1.7.32
Hangfire.SqlServer v1.6.6 to v1.7.32 (required)
Hangfire.PostgreSql v1.4.5 to v1.9.9

Hangfire Postgres schema was version 7, and it attempts to upgrade to version 17

**Can we temporarily stop the migration attempt by setting the PrepareSchemaIfNecessary flag to false in the PostgreSqlStorageOptions?**

It looks like we should manually upgrade the hangfire database ourselves, similar to this for SqlServer:
https://docs.hangfire.io/en/latest/upgrade-guides/upgrading-to-hangfire-1.7.html#migrating-the-schema
**Do you have anything equivalent to that we should be using, or suggestions on how to perform the upgrade in the production environment?**
It is vital that we retain all the current jobs and recurring jobs, and not just blow away and recreate the schema.

Thanks in advance

--

Here's the error logs:

Start installing Hangfire SQL objects...

Npgsql.NpgsqlException (0x80004005): Exception while reading from stream ---> System.TimeoutException: Timeout during reading attempt
   at Npgsql.Internal.NpgsqlReadBuffer.<<Ensure>g__EnsureLong|41_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Npgsql.Internal.NpgsqlConnector.<<ReadMessage>g__ReadMessageLong|211_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Npgsql.NpgsqlDataReader.<NextResult>d__47.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Npgsql.NpgsqlDataReader.NextResult()
   at Npgsql.NpgsqlCommand.<ExecuteReader>d__116.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Npgsql.NpgsqlCommand.<ExecuteReader>d__116.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Npgsql.NpgsqlCommand.<ExecuteNonQuery>d__104.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Npgsql.NpgsqlCommand.ExecuteNonQuery()
   at Hangfire.PostgreSql.PostgreSqlObjectsInstaller.Install(NpgsqlConnection connection, String schemaName)

Then, 9 times this error
System.InvalidOperationException: Connection is not open
   at Npgsql.NpgsqlCommand.CheckAndGetConnection()
   at Npgsql.NpgsqlCommand.<ExecuteReader>d__116.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Npgsql.NpgsqlCommand.<ExecuteScalar>d__107.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Npgsql.NpgsqlCommand.ExecuteScalar()
   at Hangfire.PostgreSql.PostgreSqlObjectsInstaller.VersionAlreadyApplied(NpgsqlConnection connection, String schemaName, Int32 version)
   at Hangfire.PostgreSql.PostgreSqlObjectsInstaller.Install(NpgsqlConnection connection, String schemaName)
   
Hangfire SQL objects installed.

**Comment by azygis** [comment]
Looks like 120s is not enough for the migration to go through.

One way to do that quicker is manually updating the schema yourself and setting the version in the schema table. Simply copy the Install.v17.sql content to pgAdmin, HeidiSQL or anywhere you can execute it, make sure the schema name matches your schema name, execute it and then change the `schema.version` to 17. After that you could try automatic migrations again.

We should make it possible to have a configurable command timeout for migrations, for sure, but unfortunately you would need to upgrade to the latest version of the package. From the issue it seems like you upgraded to the older version than is currently available - is there a reason for that?

**Comment by pjninnis** [comment]
Thanks for the quick response.

The schema remains at version 7 and hangfire is running ok against that, with our latest application code, so it looks like the state is currently valid. I assume that setting PrepareSchemaIfNeccessary to false will stop any further migration attempts, in preparation for a manual upgrade?

To confirm: to manually upgrade, I assume we'd need to apply each of Install.v8.sql upto and including Install.v17.sql in turn, and then set the schema.version to 17, correct?

I think that we upgraded only to Hangfire.PostgreSQL 1.9.9 because that was the latest at the time the upgrade was commenced. We'll be updating further with the upgrade of our app to ASP.NET Core/ .NET 6

Thanks again

**Comment by azygis** [comment]
> The schema remains at version 7 and hangfire is running ok against that, with our latest application code, so it looks like the state is currently valid.

Hmm, in that case it's already v8 which is failing. It kinda explains why there are 9 other "Connection is not open" errors. The state of schema is still invalid for the 1.9.9 version of storage, you were just lucky that you did not experience any exceptions because of schema differences _yet_.

> I assume that setting PrepareSchemaIfNeccessary to false will stop any further migration attempts, in preparation for a manual upgrade?

That is correct.

> To confirm: to manually upgrade, I assume we'd need to apply each of Install.v8.sql upto and including Install.v17.sql in turn, and then set the schema.version to 17, correct?

Based on the new information that v8 is failing, then yes, you can manually run all the scripts. You can even do all of them in a single transaction, and set the new schema version at the end of it. And since you're going manual, you should run all scripts up to and including the last one, which was at the time of v1.9.9 git tag. Technically you could also just migrate v8 script manually, and then let the startup migration try again, maybe it would work afterwards, but it might be too inconvenient to wait and see depending on your deployment strategies. At this point it's wise to just migrate yourself without 120s command timeout.

**Comment by pjninnis** [comment]
Thanks, we've upgraded manually using those scripts from schema version 7 to 17.

The initial ALTER TABLE "counter" ALTER COLUMN value TYPE bigint took 16 mins to update, due to it having 74 million row, that's why the automatic migration was failing after 2 mins.
I'm not sure why the update table is that big.

The Hangfire Jobs are successfully running again, but we are now seeing the number of processes trying to update the counter table, and they are increasing .. looks like they're trying to cull the rows and its impacting the database and timing out again.

The Processes are all attempting the same type of query:
```
            WITH "counters" AS (
              DELETE FROM "hangfire"."counter"
              WHERE "key" = $1
              AND "expireat" IS NULL
              RETURNING *
            )

            SELECT SUM("value") FROM "counters"
```

**Comment by pjninnis** [comment]
It looks like we have 72+ million rows in the counter table with key = 'stats:succeeded' and expireat is null.
If that's what its trying to delete, I suspect that's what is timing out? Should we manually delete them or are they needed?

There's also 74K rows with key = 'stats:deketed' and expireat is null.

**Comment by pjninnis** [comment]
Azygis,
    Thanks for all the help.
Yes, the major issues all came down to the size of counters table.
We truncated the table (no problems with it resetting the statistics), and everything is running smoothly (much smoother in fact) now.

Note: the automatic upgrade worked without a hitch from v7 -> v17 for our same app in a different region when we truncated the table prior to the deployment (upgraded in much less than the timeout of 2 mins).

We'll upgrade further soon, and then it will be using the aggregates then which will avoid future problems.
